FrameMetadata/Lets-Build-OTPublisher/TBExampleVideoCapture.m

```diff
@@ -0,0 +1,891 @@
+            bytesCopied += planeImageWidth;
+        }
+    }
+    assert(bytesCopied == sanitaryBufferSize);
+    *data = newImageBuffer;
+    return bytesCopied;
+}
+
+- (void)captureOutput:(AVCaptureOutput *)captureOutput
+didOutputSampleBuffer:(CMSampleBufferRef)sampleBuffer
+       fromConnection:(AVCaptureConnection *)connection {
+    
+    if (!(_capturing && _videoCaptureConsumer)) {
+        return;
+    }
+    
+    // attach time stamp to a frame
+    
+    CMTime time = CMSampleBufferGetPresentationTimeStamp(sampleBuffer);
+    CVImageBufferRef imageBuffer = CMSampleBufferGetImageBuffer(sampleBuffer);
+    
+    CVPixelBufferLockBaseAddress(imageBuffer, 0);
+    
+#warning custom method called here
+//    [self setCountToVideoFrame:_videoFrame];
+    
+    [self setTimestampToVideoFrame:_videoFrame];
+    
+//    [self setLocationToVideoFrame:_videoFrame];
+    
+//    CIImage *ciImage = [[CIImage alloc] initWithCVPixelBuffer:imageBuffer];
+//    [self setFaceDetectionTo:_videoFrame
+//                   fromImage:ciImage];
+    
+    _videoFrame.timestamp = time;
+    uint32_t height = (uint32_t)CVPixelBufferGetHeight(imageBuffer);
+    uint32_t width = (uint32_t)CVPixelBufferGetWidth(imageBuffer);
+    if (width != _captureWidth || height != _captureHeight) {
+        [self updateCaptureFormatWithWidth:width height:height];
+    }
+    _videoFrame.format.imageWidth = width;
+    _videoFrame.format.imageHeight = height;
+    CMTime minFrameDuration;
+    if (SYSTEM_VERSION_GREATER_THAN_OR_EQUAL_TO(@"7.0")) {
+        minFrameDuration = _videoInput.device.activeVideoMinFrameDuration;
+    } else {
+        AVCaptureConnection *conn =
+        [_videoOutput connectionWithMediaType:AVMediaTypeVideo];
+        minFrameDuration = conn.videoMinFrameDuration;
+    }
+    _videoFrame.format.estimatedFramesPerSecond =
+    minFrameDuration.timescale / minFrameDuration.value;
+    // TODO: how do we measure this from AVFoundation?
+    _videoFrame.format.estimatedCaptureDelay = 100;
+    _videoFrame.orientation = [self currentDeviceOrientation];
+    
+    [_videoFrame clearPlanes];
+    uint8_t* sanitizedImageBuffer = NULL;
+    
+    if (!CVPixelBufferIsPlanar(imageBuffer))
+    {
+        [_videoFrame.planes
+         addPointer:CVPixelBufferGetBaseAddress(imageBuffer)];
+    } else if ([self imageBufferIsSanitary:imageBuffer]) {
+        for (int i = 0; i < CVPixelBufferGetPlaneCount(imageBuffer); i++) {
+            [_videoFrame.planes addPointer:
+             CVPixelBufferGetBaseAddressOfPlane(imageBuffer, i)];
+        }
+    } else {
+        [self sanitizeImageBuffer:imageBuffer
+                             data:&sanitizedImageBuffer
+                           planes:_videoFrame.planes];
+    }
+    
+    [_videoCaptureConsumer consumeFrame:_videoFrame];
+    
+    free(sanitizedImageBuffer);
+    
+    CVPixelBufferUnlockBaseAddress(imageBuffer, 0);
+    
+}
+
+#warning custom set frame metadata method
```

Comment:
@robjperez assign metadata here.

Reply:
I would not mix things in the sample. I think it is better to just show one feature. We can create several capturers (`TBExampleCapturer` subclasses) that provide different functionality but do not mix everything in a single class